### DISCUSSION TITLE
`triu_indices` in JAX quite slow

### DISCUSSION THREAD

**Discussion by soraros** [jax-ml/jax#14336] [General] [answered]
To avoid materialising the output during staging, I wrote a version of `triu_indices` in JAX. I think I did a ok job in using the `numpy`-style array DSL. However, even the `jit`ted version is quite slow (at least on cpu) comparing with `np.triu_indices`. Could someone help understand why it is the case? Is it because, for instance, `cumsum` is not "entirely vectorized"?

```python
@partial(jit, static_argnames=('n', 'k'))
def triu_indices(n: int, k: int = 0):
  assert n >= 0 and k >= 0
  N = max(n - k, 0)
  l = N * (N + 1) // 2
  iota = lax.iota(np.int32, N)
  idx = lax.cumsum(1 + iota, reverse=True)
  I = lax.cumsum(jnp.zeros(l, dtype=np.int32).at[idx].set(1))
  J = lax.iota(np.int32, l) - lax.cumsum(jnp.zeros(l, dtype=np.int32).at[idx].set(iota))
  return I, J + k
```

**Comment by jakevdp** [comment, marked as answer]
When I try your code on a Colab CPU runtime, I find that your JAX version is pretty fast. Can you say more about how you benchmarked this?
```python
_ = jax.block_until_ready(triu_indices(10))
%timeit jax.block_until_ready(triu_indices(10))
# 8.37 µs ± 79.3 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

%timeit np.triu_indices(10)
# 23.3 µs ± 4.3 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

**Reply by soraros** [reply]
It doesn't look as pretty when `n` is larger, say 1000 or 1e4. This is a bit surprising as I'd expect the JAX version to scale better than its `numpy` counterpart.

<details>
<summary>Benchmark script here</summary>

```python

from functools import partial

import numpy as np

import jax
import jax.numpy as jnp
from jax import jit, lax

@partial(jit, static_argnames=('n', 'k'))
def triu_indices(n: int, k: int = 0):
  assert n >= 0 and k >= 0
  N = max(n - k, 0)
  l = N * (N + 1) // 2
  iota = lax.iota(np.int32, N)
  idx = lax.cumsum(1 + iota, reverse=True)
  I = lax.cumsum(jnp.zeros(l, dtype=np.int32).at[idx].set(1))
  J = lax.iota(np.int32, l) - lax.cumsum(jnp.zeros(l, dtype=np.int32).at[idx].set(iota))
  return I, J + k


n = 10_000
k = 1

%timeit np.triu_indices(n, k)
# 468 ms ± 41 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

_ = jax.block_until_ready(triu_indices(n, k))
%timeit jax.block_until_ready(triu_indices(n, k))
# 1.62 s ± 302 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

</details>

**Reply by jakevdp** [reply]
Ah, thanks, that makes sense. I wonder if the problem is that your approach relies on multiple scatter and cumsum operations, while numpy's implementation is more direct. Maybe something like this would be more performant?
```python
@partial(jax.jit, static_argnames=['n', 'k'])
def triu_indices(n, k=0):
  assert 0 <= k <= n  # logic below needs to be modified to handle other values
  size = (n - k) * (n - k + 1) // 2
  idx = jnp.arange(n)
  return jnp.where(idx >= idx[:, None] + k, size=size)
```
It uses about 2x the memory, but relies on more efficient operations.

**Reply by soraros** [reply]
> relies on multiple scatter and cumsum operations

I figured as much. But I can see some other operations (say `bincount`) implemented in the same fashion and they all seem to run pretty fast.

> It uses about 2x the memory, but relies on more efficient operations.

Yeah, I'm tried something like this (`jnp.nonzero(jnp.tri(...))`) as well, but it seems to be even slower.

I should add on GPU, then `scatter` version runs the fastest, and way faster than `numpy`.

**Reply by jakevdp** [reply]
Thanks - it's not surprising that this isn't particularly performant on CPU as compared to GPU, because there hasn't been a lot of work in optimizing XLA CPU performance, and by contrast there's been decades of work in optimiizing numpy's CPU performance. This is discussed in broad strokes at https://jax.readthedocs.io/en/latest/faq.html#is-jax-faster-than-numpy

**Reply by jakevdp** [reply]
> But I can see some other operations (say bincount) implemented in the same fashion and they all seem to run pretty fast.

Have you benchmarked them? These operations were not implemented this way for performance, but rather because they're just about the only way to implement such things on top of the primitives provided by XLA.

**Reply by soraros** [reply]
> but rather because they're just about the only way to implement such things on top of the primitives provided by XLA

This I'm aware of.

> Have you benchmarked them?

Good point. I suppose I'm under the (false?) impression that they (`histogram`, `bincount` etc.) are faster because part of the overhead is hidden in jit-compiled code. Or, maybe I'm just too used to seeing JAX run way faster than `numpy` even on CPU due of fusion, DCE etc.

Regarding `triu_indices`, I can't think of an implementation with drastically better performance in terms of complexity than the two mentioned in our discussion. Does this mean we can't have a `triu_indices` provided by `jax.numpy` which doesn't eagerly materialise its output in the near future?

**Reply by jakevdp** [reply]
We can have this kind of implementation (and we probably should) but it will be slower than NumPy on CPU. I don't think that's a deal-breaker; I think it's pretty typical of the performance of JAX wrappers of nontrivial numpy functions.

**Reply by soraros** [reply]
> We can have this kind of implementation (and we probably should) but it will be slower than NumPy on CPU. I don't think that's a deal-breaker; I think it's pretty typical of the performance of JAX wrappers of nontrivial numpy functions.

Totally agree. I opened #14339 to track this.

**Reply by jakevdp** [reply]
Thanks!
